Fix setting fog default values when calling setStyle

avpeery opened the pull request:
This fixes #11978 which removes a previous [change](https://github.com/mapbox/mapbox-gl-js/blob/59e3d813ecf465e6869e6184bb9154f8dfd653b1/src/style/fog.js#L84) that was added to use the old fog style spec properties when toggling if it wasn't set in the new style spec rather than using the default values. 

**Before:**

https://user-images.githubusercontent.com/42715836/173741641-9de4be41-7401-4fe0-8d29-0fc1f9a1a47a.mov


**After:**

https://user-images.githubusercontent.com/42715836/173741617-7df81188-7cab-4eb2-b7a3-d7222b4c9326.mov


## Launch Checklist

<!-- Thanks for the PR! Feel free to add or remove items from the checklist. -->

 - [x] briefly describe the changes in this PR
 - [x] include before/after visuals or gifs if this PR includes visual changes
 - [x] write tests for all new functionality
 - [x] manually test the debug page
 - [x] add an entry inside this element for inclusion in the `mapbox-gl-js` changelog: `<changelog>Allow style spec to use fog default values when toggling styles </changelog>`

karimnaaji commented:
@avpeery for render tests you can remove the render tests skipped from `test/ignores.json` that were added in https://github.com/mapbox/mapbox-gl-js/pull/11990:

```
  "render-tests/fog/switch-style-disable": "skip - not implemented yet",
  "render-tests/fog/switch-style-reset-defaults": "skip - not implemented yet"
```
as these two tests should now pass and will confirm we have the same behavior as with native.

karimnaaji reviewed (commented):
The change looks good to me, any idea why the render tests are failing?

avpeery commented:
> The change looks good to me, any idea why the render tests are failing?

Was focusing on a different fix to allow `map.setFog()` respecting old property values while `setStyle` respects default when toggling. This PR will prefer default values in both cases (which is better than what currently is in main for studio). I'll link the other PR when opened, but still some failing tests. The reason the render test was failing here was I deleted the wrong render test from ignores.json, and fixed it now! This could be a good option to merge in until a good solution is done to allow the other functionality for map.setFog

SnailBones reviewed (approved):
Code looks good to me!

Tests pass for me after adding a `wait` at the end of `switch-style-reset-defaults` (as you have done with `switch-style-disable`)

<img width="1111" alt="image" src="https://user-images.githubusercontent.com/14878684/174345909-fcde779b-35ab-4723-ba61-9a1b782d2c0f.png">

I'm noticing that calling `setFog()` removes the fog instead of setting it back to the style sheet ([the behavior of prjoections](https://github.com/mapbox/mapbox-gl-js/pull/11204)). I'm not sure what the best behavior is here but consistency would be nice, perhaps this could be part of of future style refactor?

karimnaaji commented on test/ignores.json:
```diff
@@ -40,6 +40,5 @@
   "render-tests/globe/globe-transition/horizon-during-transition": "skip - pending globe-view implementation",
-  "render-tests/fog/switch-style-disable": "skip - not implemented yet",
-  "render-tests/fog/switch-style-reset-defaults": "skip - not implemented yet"
+  "render-tests/fog/switch-style-disable": "skip - not implemented yet"
```
```suggestion
```

karimnaaji commented on src/style/style.js:
```diff
@@ -1554,9 +1554,13 @@ class Style extends Evented {
             const fog = this.fog;
             const currSpec = fog.get();
+
+            // empty object should pass through to set default values
+            if (Object.keys(fogOptions).length === 0) fog.set(fogOptions);
```
Could we have a render test for this code path? Based on our conversation, it'd be something like:
```
map.setFog({
    'color': '#0F2127',
    'high-color': '#000',
    'horizon-blend': 0.5,
    'space-color': '#000'
})
map.setFog({})
```

avpeery replied:
Ok sounds good, Yes quickly confirmed that currently in this use-case, the fog isn't changed on the map.setFog({}) call

karimnaaji reviewed (approved):
:+1:
